cluster-api/machines-api/types.go

```diff
@@ -0,0 +1,160 @@
+package machines
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/api"
+)
+
+// Machine represents a single Node that should exist (whether it does or
+// not yet). In this model, there is no grouping of nodes to scale with a
+// numeric field. Each Machine exists independently, and grouping can only
+// be inferred via label selectors.
+//
+// In order for a new Node to be created, one can generically create a new
+// Machine object, possibly copying the spec from an existing Machine
+// or a template. To scale down the cluster, delete specific instances of
+// Machines and the underlying Nodes will be unregistered/deprovisioned.
+// Separate provider-specific controllers will watch Machine objects they can
+// act on (like a GCE cloud controller watching for only Machines destined for
+// GCE) and take the appropriate actions.
+//
+// Any updates to the MachineSpec will be actuated to change the Node in
+// place or replace the Node with one conforming to the spec. In this model,
+// the fact that a controller is able to upgrade a node via in-place upgrades
+// or via a cloud replacement is an implementation detail without API controls.
+//
+// It is recommended, but not required, that provider-specific controllers add
+// finalizers to Machine objects so that they can be triggered on deletion to
+// release the necessary external resources, reporting any errors encountered.
+type Machine struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec   MachineSpec
+	Status MachineStatus
+}
+
+type MachineSpec struct {
+	// This ObjectMeta will autopopulate the Node created. Use this to
+	// indicate what labels, annotations, name prefix, etc., should be used
```

Comment:
Actually, I was aiming to follow the same pattern as we do for pods (and other objects): you can specify `name: value` if you know exactly what name you want to use, or `generateName: value` to use `value` as a prefix and have a unique suffix generated for you.

I can call this out much more explicitly, though.

Reply:
Isn't `generateName` handled generically by the API server? How will you prevent it from renaming the `MachineSpec` you create, so that providers can generate the names (e.g. GKE has a way of generating names that is *not* the same as what the API server does)?

OTOH you could just force providers to use the name generated by the API server. Though IDK what kinds of incompatibility that would introduce.

Keep in mind that using `generateName` will prevent you from making idempotent creation requests, because it is not deterministic.